jumanji/environments/routing/mmst/env.py

```diff
@@ -0,0 +1,762 @@
+            step_key, jnp.arange(self.num_agents)
+        )
+
+        def not_all_agents_actions_examined(arg: Any) -> Any:
+            added_nodes, new_actions, action, nodes, agent_permutation, index = arg
+            return index < self.num_agents
+
+        def modify_action_if_agent_target_node_is_selected(arg: Any) -> Any:
+            added_nodes, new_actions, action, nodes, agent_permutation, index = arg
+            agent_i = agent_permutation[index]
+
+            is_invalid_node = nodes[agent_i] == EMPTY_NODE
+            node_is_not_selected = jnp.sum(jnp.sum(added_nodes == nodes[agent_i]) == 0)
+
+            # false + false = 0 = tie break
+            # true + false = 1  = invalid choice (with tie break) do nothing
+            # false + true * 2 = 2 = valid choice and valid node
+            # true + true * 2 = 3 -> invalid choice (without tie break)
+
+            new_actions, added_nodes = jax.lax.switch(
+                is_invalid_node + node_is_not_selected * 2,
+                [
+                    lambda *_: (
+                        new_actions.at[agent_i].set(INVALID_TIE_BREAK),
+                        added_nodes.at[agent_i].set(INVALID_TIE_BREAK),
+                    ),
+                    lambda *_: (new_actions, added_nodes),
+                    lambda *_: (
+                        new_actions.at[agent_i].set(action[agent_i]),
+                        added_nodes.at[agent_i].set(nodes[agent_i]),
+                    ),
+                    lambda *_: (new_actions, added_nodes),
+                ],
+                new_actions,
+                added_nodes,
+            )
+            index += 1
+
+            return (added_nodes, new_actions, action, nodes, agent_permutation, index)
+
+        (
+            added_nodes,
+            new_actions,
+            action,
+            nodes,
+            agent_permutation,
+            index,
+        ) = jax.lax.while_loop(
+            not_all_agents_actions_examined,
+            modify_action_if_agent_target_node_is_selected,
+            (added_nodes, new_actions, action, nodes, agent_permutation, 0),
+        )
+
+        def mask_visited_nodes(
+            node_visited: jnp.int32, oldaction: jnp.int32
+        ) -> jnp.int32:
+            new_action = jax.lax.cond(  # type:ignore
+                node_visited != EMPTY_NODE,
+                lambda *_: INVALID_ALREADY_TRAVERSED,
+                lambda *_: oldaction,
+            )
+
+            return new_action
+
+        final_actions = jnp.zeros_like(new_actions)
+        # set the action to 0 if the agent is moving to an already connected node
+        for agent in range(self.num_agents):
+            node_visited = state.connected_nodes_index[agent, nodes[agent]]
+            new_action = mask_visited_nodes(node_visited, new_actions[agent])
+            final_actions = final_actions.at[agent].set(new_action)
+
+        # masked agents with finished states
```

Comment:
```suggestion
        # Masked agents with finished states.
```
Please can you review and make sure comments are proper sentences?

Reply:
Sure I will go through all the comments in the environment now.